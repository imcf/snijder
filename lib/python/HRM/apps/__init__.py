# -*- coding: utf-8 -*-
"""
GC3lib based abstract application class.

Classes
-------

AbstractApp()
"""

import os
import gc3libs

from .. import logi, logd, logw, logc, loge


class AbstractApp(gc3libs.Application):

    """App object for generic gc3lib based jobs.

    This virtual application is to be used for deriving subclasses that share
    the common methods defined here.
    """

    def __init__(self, job, appconfig):
        """Set up the application.

        Parameters
        ----------
        job : HRM.jobs.JobDescription
        appconfig : dict
            A dict with at least all mandatory parameters for a
            gc3libs.Application, plus possibly extra parameters.
        """
        if self.__class__.__name__ == 'AbstractApp':
            raise TypeError("Refusing to instantiate class 'AbstractApp'!")
        self.job = job   # remember the job object
        logd('gc3_output_dir: %s', appconfig['output_dir'])
        logd('self.job: %s', job)
        logw('Instantiating a %s: [[user: %s]] [[uid: %s]]',
             self.__class__.__name__, job['user'], job['uid'])
        super(AbstractApp, self).__init__(**appconfig)
        self.laststate = self.execution.state

    def new(self):
        """Called when the job state is (re)set to NEW.

        Note this will not be called when the application object is created,
        rather if the state is reset to NEW after it has already been
        submitted.
        """
        self.status_changed()

    def running(self):
        """Called when the job state transitions to RUNNING."""
        self.status_changed()

    def stopped(self):
        """Called when the job state transitions to STOPPED."""
        self.status_changed()
        logc("Job '%s' has been suspended for an unknown reason!!!",
             self.job['uid'])

    def submitted(self):
        """Called when the job state transitions to SUBMITTED."""
        self.status_changed()

    def terminated(self):
        """This is called when the app has terminated execution."""
        self.status_changed()
        if self.execution.exitcode is None:
            # TODO: we could let the app know it was killed
            #       currently, were guessing from the exitcode 'None' that the
            #       app was explicitly killed by gc3pie - it would be cleaner
            #       to explicitly cover this situation e.g. in the spooler's
            #       cleanup() method by telling the app it is requested to stop
            logw("Job '%s' apparently was killed or crahsed!", self.job['uid'])
        elif self.execution.exitcode != 0:
            # IMPORTANT: gc3pie does NOT seem to pass on the exit code of
            # the job in this value, instead every non-zero exit code is
            # represented as 255 - which means we can NOT DERIVE from this how
            # the job process has finished!
            logc("Job '%s' terminated with unexpected EXIT CODE: %s!",
                 self.job['uid'], self.execution.exitcode)
        else:
            logi("Job '%s' terminated successfully!", self.job['uid'])
            logi("The output of the application is in `%s`.", self.output_dir)

    def terminating(self):
        """Called when the job state transitions to TERMINATING."""
        self.status_changed()

    def status_changed(self):
        """Check the if the execution state of the app has changed.

        Track and update the internal execution status of the app and print a
        log message if the status changes. Return the new state if the app it
        has changed, otherwise None.
        """
        new = self.execution.state
        if new != self.laststate:
            logi("%s: job status changed from '%s' to '%s'.",
                 self.__class__.__name__, self.laststate, new)
            self.laststate = self.job['status'] = new
            return new
        else:
            return None
