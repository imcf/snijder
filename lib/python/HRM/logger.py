#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
Logging helper module.
"""

import logging
import gc3libs

__all__ = ["logw", "logi", "logd", "loge", "logc"]

# we set a default loglevel and add some shortcuts for logging:
LOGLEVEL = logging.WARN
gc3libs.configure_logger(LOGLEVEL, "qmgc3")

logw = gc3libs.log.warn
logi = gc3libs.log.info
logd = gc3libs.log.debug
loge = gc3libs.log.error
logc = gc3libs.log.critical
