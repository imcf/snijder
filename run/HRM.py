#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
Support for various HRM related tasks.

Classes
-------

JobDescription()
    Parser for job descriptions, works on files or strings.

JobQueue()
    Job handling and scheduling.
"""

import ConfigParser
import pprint
from collections import deque
from hashlib import sha1

from hrm_logger import warn, info, debug, set_loglevel

__all__ = ['JobDescription', 'JobQueue']


# expected version for job description files:
JOBFILE_VER = '3'

def deq_del(d, n):
    """Delete the n'th element of a deque object."""
    d.rotate(-n)
    d.popleft()
    d.rotate(n)


class JobDescription(dict):

    """Abstraction class for handling HRM job descriptions.

    Read an HRM job description either from a file or a string and parse
    the sections, check them for sane values and store them in a dict.
    """

    def __init__(self, job, srctype):
        """Initialize depending on the type of description source.

        Parameters
        ----------
        job : string
        srctype : string

        Example
        -------
        >>> job = HRM.JobDescription('/path/to/jobdescription.cfg', 'file')
        """
        self.jobparser = ConfigParser.RawConfigParser()
        self._sections = []
        if (srctype == 'file'):
            self.name = "file '%s'" % job
            self._parse_jobfile(job)
        elif (srctype == 'string'):
            # TODO: _parse_jobstring(job)
            self.name = "string received from socket"
            raise Exception("Source type 'string' not yet implemented!")
        else:
            raise Exception("Unknown source type '%s'" % srctype)
        # store the SHA1 digest of this job, serving as the UID:
        # TODO: this should better be the hash of the actual (unparsed) string
        # instead of the representation of the Python object, but therefore
        # we need to hook into the parsing itself (or read the file twice).
        self['uid'] = sha1(self.__repr__()).hexdigest()
        pprint.pprint("Finished initialization of JobDescription().")
        pprint.pprint(self)

    def _parse_jobfile(self, fname):
        """Initialize ConfigParser for a file and run parsing method."""
        self.jobparser.read(fname)
        self._parse_jobdescription()

    def _parse_jobdescription(self):
        """Parse details for an HRM job and check for sanity.

        Use the ConfigParser object and assemble a dicitonary with the
        collected details that contains all the information for launching a new
        processing task. Raises Exceptions in case something unexpected is
        found in the given file.
        """
        # FIXME: currently only deconvolution jobs are supported, until hucore
        # will be able to do the other things like SNR estimation and
        # previewgen using templates as well!
        # TODO: group code into parsing and sanity-checking
        self._sections = self.jobparser.sections()
        # parse generic information, version, user etc.
        if not 'hrmjobfile' in self._sections:
            raise ValueError("Error parsing job from %s." % self.name)
        try:
            self['ver'] = self.jobparser.get('hrmjobfile', 'version')
        except ConfigParser.NoOptionError:
            raise ValueError("Can't find version in %s." % self.name)
        if not (self['ver'] == JOBFILE_VER):
            raise ValueError("Unexpected version in %s." % self['ver'])
        try:
            self['user'] = self.jobparser.get('hrmjobfile', 'username')
        except ConfigParser.NoOptionError:
            raise ValueError("Can't find username in %s." % self.name)
        try:
            self['email'] = self.jobparser.get('hrmjobfile', 'useremail')
        except ConfigParser.NoOptionError:
            raise ValueError("Can't find email address in %s." % self.name)
        try:
            self['type'] = self.jobparser.get('hrmjobfile', 'jobtype')
        except ConfigParser.NoOptionError:
            raise ValueError("Can't find jobtype in %s." % self.name)
        # from here on a jobtype specific parsing must be done:
        if self['type'] == 'hucore':
            self._parse_job_hucore()
        else:
            raise ValueError("Unknown jobtype '%s'" % self['type'])

    def _parse_job_hucore(self):
        """Do the specific parsing of "hucore" type jobfiles.

        Parse the "hucore" and the "inputfiles" sections of HRM job
        configuration files.

        Returns
        -------
        void
            All information is added to the "self" dict.
        """
        # the "hucore" section:
        try:
            self['exec'] = self.jobparser.get('hucore', 'executable')
        except ConfigParser.NoOptionError:
            raise ValueError("Can't find executable in %s." % self.name)
        try:
            self['template'] = self.jobparser.get('hucore', 'template')
        except ConfigParser.NoOptionError:
            raise ValueError("Can't find template in %s." % self.name)
        # and the input file(s):
        if not 'inputfiles' in self._sections:
            raise ValueError("No input files defined in %s." % self.name)
        self['infiles'] = []
        for option in self.jobparser.options('inputfiles'):
            infile = self.jobparser.get('inputfiles', option)
            self['infiles'].append(infile)

    def get_category(self):
        """Get the category of this job, in our case the value of 'user'."""
        return self['user']


class JobQueue(object):

    """Class to store a list of jobs that need to be processed.

    An instance of this class can be used to keep track of lists of jobs of
    different categories (e.g. individual users). The instance will contain a
    scheduler so that it is possible for the caller to simply request the next
    job from this queue without having to care about priorities or anything
    else.
    """

    # TODO: implement len() or similar
    # TODO: either remove items from jobs[] upon pop() / remove() or add their
    # ID to a list so the jobs[] dict can get garbage-collected later
    def __init__(self):
        """Initialize an empty job queue."""
        self.cats = deque('')  # categories / users, used by the scheduler
        # jobs is a dict containing the JobDescription objects using their
        # UID as the indexing key for fast access:
        self.jobs = dict()
        self.queue = dict()

    def append(self, job):
        """Add a new job to the queue."""
        # TODO: should we catch duplicate jobs? Currently they are enqueued.
        cat = job.get_category()
        uid = job['uid']
        info("Enqueueing job '%s' into category '%s'." % (uid, cat))
        self.jobs[uid] = job  # store the job in the global dict
        if not cat in self.cats:
            warn("Adding a new queue for '%s' to the JobQueue." % cat)
            self.cats.append(cat)
            self.queue[cat] = deque()
            debug("Current queue categories: %s" % self.cats)
        else:
            # in case there are already jobs of this category, we don't touch
            # the scheduler / priority queue:
            debug("JobQueue already contains a queue for '%s'." % cat)
        self.queue[cat].append(uid)
        info("Queue for category '%s': %s" % (cat, self.queue[cat]))
        # debug("Overall list of job descriptions: %s" % self.jobs)

    def pop(self):
        """Return the next job description for processing."""
        # TODO: catch an empty categories queue
        cat = self.cats[0]
        jobid = self.queue[cat].popleft()
        info("Retrieving next job: category '%s', uid '%s'." % (cat, jobid))
        if len(self.queue[cat]) >= 1:
            debug("Shifting category list.")
            self.cats.rotate(-1)  # move the first element to last position
        else:
            debug("Queue for category '%s' now empty, removing it." % cat)
            self.cats.popleft()  # remove it from the categories list
            del self.queue[cat]  # delete the category from the queue dict
        debug("Current queue categories: %s" % self.cats)
        debug("Current contents of all queues: %s" % self.queue)
        return self.jobs[jobid]

    def remove(self, uid):
        """Remove a job with a given UID from the queue."""
        # TODO: cover non-existing UID's
        warn("Trying to remove job with uid '%s'." % uid)
        cat = self.jobs[uid].get_category()
        debug("Category of job to remove: '%s'." % cat)
        self.queue[cat].remove(uid)
        debug("Current queue categories: %s" % self.cats)
        debug("Current contents of all queues: %s" % self.queue)
        if len(self.queue[cat]) < 1:
            debug("Queue for category '%s' now empty, removing it." % cat)
            self.cats.remove(cat)  # remove it from the categories list
            del self.queue[cat]    # delete the category from the queue dict
            debug("Current queue categories: %s" % self.cats)
            debug("Current contents of all queues: %s" % self.queue)
