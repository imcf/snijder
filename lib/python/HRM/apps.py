#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
Support for various HRM related tasks, GC3lib application classes.

Classes
-------

HucoreDeconvolveApp()
HucorePreviewgenApp()
HucoreEstimateSNRApp()
    The gc3libs applications.
"""

from .logger import *

import os
import gc3libs

class HucoreDeconvolveApp(gc3libs.Application):

    """App object for 'hucore' deconvolution jobs.

    This application calls `hucore` with a given template file and retrives the
    stdout/stderr in a file named `stdout.txt` plus the directories `resultdir`
    and `previews` into a directory `deconvolved` inside the current directory.
    """

    def __init__(self, job, gc3_output):
        self.job = job   # remember the job object
        uid = self.job['uid']
        logw('Instantiating a HucoreDeconvolveApp:\n[%s]: %s --> %s',
             self.job['user'], self.job['template'], self.job['infiles'])
        logi('Job UID: %s', uid)
        # we need to add the template (with the local path) to the list of
        # files that need to be transferred to the system running hucore:
        self.job['infiles'].append(self.job['template'])
        # for the execution on the remote host, we need to strip all paths from
        # this string as the template file will end up in the temporary
        # processing directory together with all the images:
        templ_on_tgt = self.job['template'].split('/')[-1]
        gc3libs.Application.__init__(
            self,
            arguments=[self.job['exec'],
                       '-exitOnDone',
                       '-noExecLog',
                       '-checkForUpdates', 'disable',
                       '-template', templ_on_tgt],
            inputs=self.job['infiles'],
            outputs=['resultdir', 'previews'],
            # collect the results in a subfolder of GC3Pie's spooldir:
            output_dir=os.path.join(gc3_output, 'results_%s' % uid),
            stderr='stdout.txt', # combine stdout & stderr
            stdout='stdout.txt'
        )
        self.laststate = self.execution.state

    def terminated(self):
        """This is called when the app has terminated execution."""
        # TODO: put the results dir back to the user's destination directory
        # (WARNING: we have to be careful if the data has already been
        # collected in case of a remote execution scenario)
        # TODO: consider specifying the output dir in the jobfile!
        # -> for now we simply use the gc3spooldir as the output directory to
        # ensure results won't get moved across different storage locations:
        # hucore EXIT CODES:
        # 0: all went well
        # 143: hucore.bin received the HUP signal (9)
        # 165: the .hgsb file could not be parsed (file missing or with errors)
        if self.execution.exitcode != 0:
            logc("Job '%s' terminated with unexpected EXIT CODE: %s!",
                 self.job['uid'], self.execution.exitcode)
        else:
            logi("Job '%s' terminated successfully!", self.job['uid'])
        logd("The output of the application is in `%s`.", self.output_dir)

    def status_changed(self):
        """Check the if the execution state of the app has changed.

        Track and update the internal execution status of the app and print a
        log message if the status changes. Return the new state if the app it
        has changed, otherwise None.
        """
        if self.execution.state != self.laststate:
            logi("Job status changed to '%s'.", self.job['status'])
            self.laststate = self.execution.state
            return self.execution.state
        else:
            return None


class HucorePreviewgenApp(gc3libs.Application):

    """App object for 'hucore' image preview generation jobs."""

    def __init__(self):
        # logw('Instantiating a HucorePreviewgenApp:\n%s', job)
        logw('WARNING: this is a stub, nothing is implemented yet!')
        super(HucorePreviewgenApp, self).__init__()


class HucoreEstimateSNRApp(gc3libs.Application):

    """App object for 'hucore' SNR estimation jobs."""

    def __init__(self):
        # logw('Instantiating a HucoreEstimateSNRApp:\n%s', job)
        logw('WARNING: this is a stub, nothing is implemented yet!')
        super(HucoreEstimateSNRApp, self).__init__()
