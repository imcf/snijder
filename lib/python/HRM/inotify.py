# -*- coding: utf-8 -*-
"""
Helper module for pyinotify stuff.
"""

import os
import pyinotify

from .jobs import process_jobfile
from .logger import logi, logd


class EventHandler(pyinotify.ProcessEvent):
    """Handler for pyinotify filesystem events.

    An instance of this class can be registered as a handler to pyinotify and
    then gets called to process an event registered by pyinotify.

    Public Methods
    --------------
    process_IN_CREATE()
    """

    def my_init(self, queues, dirs):                # pylint: disable=W0221
        """Initialize the inotify event handler.

        Parameters
        ----------
        queues : dict
            Containing the JobQueue objects for the different queues, using the
            corresponding 'type' keyword as identifier.
        dirs : dict
            Spooling directories in a dict, as returned by HRM.setup_rundirs().
        """
        self.queues = queues
        self.dirs = dirs
        logi('Initialized the event handler for inotify, watching job '
             'submission directory "%s".', self.dirs['new'])

    def process_IN_CREATE(self, event):
        """Method handling 'create' events.

        Parameters
        ----------
        event : pyinotify.Event
        """
        logi("New file event '%s'", os.path.basename(event.pathname))
        logd("inotify 'IN_CREATE' event full file path '%s'", event.pathname)
        process_jobfile(event.pathname, self.queues, self.dirs)