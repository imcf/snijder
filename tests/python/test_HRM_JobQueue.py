#!/usr/bin/env python

"""Simple test script for the HRM Queue Manager class.

Run it from this directory after setting your PYTHONPATH accordingly:

export PYTHONPATH=$PYTHONPATH:../../lib/python/
python test_HRM_JobQueue.py
"""

try:
    import HRM
except ImportError:
    raise SystemExit(__doc__)

# the reload statement is here so we can us the script in subsequent calls
# during a single IPython session:
reload(HRM)

jobs = list(xrange(7))

jobfile = '../jobfiles/sandbox/deconvolution_job.cfg'
for i in xrange(7):
    jobs[i] = HRM.JobDescription(jobfile, 'file')

jobs[0]['uid'] = 'u000_aaa'
jobs[0]['user'] = 'u000'
jobs[1]['uid'] = 'u000_bbb'
jobs[1]['user'] = 'u000'
jobs[2]['uid'] = 'u000_ccc'
jobs[2]['user'] = 'u000'

jobs[3]['uid'] = 'u111_ddd'
jobs[3]['user'] = 'u111'
jobs[4]['uid'] = 'u111_eee'
jobs[4]['user'] = 'u111'
jobs[5]['uid'] = 'u111_fff'
jobs[5]['user'] = 'u111'
jobs[6]['uid'] = 'u111_ggg'
jobs[6]['user'] = 'u111'

def remove_print(uid):
    job = jq.remove(uid)
    print "remove('%s'): %s (joblist: %s)" % (uid, job['uid'], jq.joblist())

def next_print():
    uid = jq.next_job()['uid']
    print "next_job(): %s (joblist: %s)" % (uid, jq.joblist())

jq = HRM.JobQueue()

print("\n******** adding jobs to queue: ********")
for job in jobs:
    jq.append(job)

print("\n******** retrieving jobs from queue: ********")
for job in jobs:
    jq.pop()


print("\n******** adding jobs to queue: ********")
for job in jobs:
    jq.append(job)

print("\n******** removing jobs from queue: ********")

remove_print('u000_aaa')
remove_print('u000_bbb')
next_print()
remove_print('u111_ggg')
next_print()
remove_print('u111_eee')
remove_print('u111_fff')
print jq.remove('aaa')
