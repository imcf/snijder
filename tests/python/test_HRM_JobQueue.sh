#!/bin/bash

export PYTHONPATH=$PYTHONPATH:../../lib/python/
python test_HRM_JobQueue.py > test_HRM_JobQueue.out

clear ; cat test_HRM_JobQueue.out

