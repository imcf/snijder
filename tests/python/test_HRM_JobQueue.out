
******** adding jobs to queue: ********
jq.joblist: []
jq.joblist: ['u000_aaa']
jq.joblist: ['u000_aaa', 'u000_bbb']
jq.joblist: ['u000_aaa', 'u000_bbb', 'u000_ccc']
jq.joblist: ['u000_aaa', 'u111_ddd', 'u000_bbb', 'u000_ccc']
jq.joblist: ['u000_aaa', 'u111_ddd', 'u000_bbb', 'u111_eee', 'u000_ccc']
jq.joblist: ['u000_aaa', 'u111_ddd', 'u000_bbb', 'u111_eee', 'u000_ccc', 'u111_fff']
jq.joblist: ['u000_aaa', 'u111_ddd', 'u000_bbb', 'u111_eee', 'u000_ccc', 'u111_fff', 'u111_ggg']
========================= queue status =========================
--- jobs retrieved for processing
None.
------------------------- queue status -------------------------
--- jobs queued (not yet retrieved)
u000 (hrm@svi.nl): u000_aaa - ['tests/jobfiles/sandbox/bad.lsm'] [queued]
u111 (hrm@svi.nl): u111_ddd - ['tests/jobfiles/sandbox/bad.lsm'] [queued]
u000 (hrm@svi.nl): u000_bbb - ['tests/jobfiles/sandbox/bad.lsm'] [queued]
u111 (hrm@svi.nl): u111_eee - ['tests/jobfiles/sandbox/bad.lsm'] [queued]
u000 (hrm@svi.nl): u000_ccc - ['tests/jobfiles/sandbox/bad.lsm'] [queued]
u111 (hrm@svi.nl): u111_fff - ['tests/jobfiles/sandbox/bad.lsm'] [queued]
u111 (hrm@svi.nl): u111_ggg - ['tests/jobfiles/sandbox/bad.lsm'] [queued]
========================= queue status =========================


******** retrieving jobs from queue for processing: ********
jq.joblist: ['u000_aaa', 'u111_ddd', 'u000_bbb', 'u111_eee', 'u000_ccc', 'u111_fff', 'u111_ggg']
next: 'u000_aaa' (joblist: ['u111_ddd', 'u000_bbb', 'u111_eee', 'u000_ccc', 'u111_fff', 'u111_ggg'])
next: 'u111_ddd' (joblist: ['u000_bbb', 'u111_eee', 'u000_ccc', 'u111_fff', 'u111_ggg'])
next: 'u000_bbb' (joblist: ['u111_eee', 'u000_ccc', 'u111_fff', 'u111_ggg'])
next: 'u111_eee' (joblist: ['u000_ccc', 'u111_fff', 'u111_ggg'])
next: 'u000_ccc' (joblist: ['u111_fff', 'u111_ggg'])
next: 'u111_fff' (joblist: ['u111_ggg'])
next: 'u111_ggg' (joblist: [])
jq.joblist: []
========================= queue status =========================
--- jobs retrieved for processing
u000 (hrm@svi.nl): u000_aaa - ['tests/jobfiles/sandbox/bad.lsm'] [queued]
u111 (hrm@svi.nl): u111_ddd - ['tests/jobfiles/sandbox/bad.lsm'] [queued]
u000 (hrm@svi.nl): u000_bbb - ['tests/jobfiles/sandbox/bad.lsm'] [queued]
u111 (hrm@svi.nl): u111_eee - ['tests/jobfiles/sandbox/bad.lsm'] [queued]
u000 (hrm@svi.nl): u000_ccc - ['tests/jobfiles/sandbox/bad.lsm'] [queued]
u111 (hrm@svi.nl): u111_fff - ['tests/jobfiles/sandbox/bad.lsm'] [queued]
u111 (hrm@svi.nl): u111_ggg - ['tests/jobfiles/sandbox/bad.lsm'] [queued]
------------------------- queue status -------------------------
--- jobs queued (not yet retrieved)
None.
========================= queue status =========================


******** adding jobs to queue: ********
jq.joblist: []
jq.joblist: ['u000_aaa']
jq.joblist: ['u000_aaa', 'u000_bbb']
jq.joblist: ['u000_aaa', 'u000_bbb', 'u000_ccc']
jq.joblist: ['u000_aaa', 'u111_ddd', 'u000_bbb', 'u000_ccc']
jq.joblist: ['u000_aaa', 'u111_ddd', 'u000_bbb', 'u111_eee', 'u000_ccc']
jq.joblist: ['u000_aaa', 'u111_ddd', 'u000_bbb', 'u111_eee', 'u000_ccc', 'u111_fff']
jq.joblist: ['u000_aaa', 'u111_ddd', 'u000_bbb', 'u111_eee', 'u000_ccc', 'u111_fff', 'u111_ggg']
========================= queue status =========================
--- jobs retrieved for processing
None.
------------------------- queue status -------------------------
--- jobs queued (not yet retrieved)
u000 (hrm@svi.nl): u000_aaa - ['tests/jobfiles/sandbox/bad.lsm'] [queued]
u111 (hrm@svi.nl): u111_ddd - ['tests/jobfiles/sandbox/bad.lsm'] [queued]
u000 (hrm@svi.nl): u000_bbb - ['tests/jobfiles/sandbox/bad.lsm'] [queued]
u111 (hrm@svi.nl): u111_eee - ['tests/jobfiles/sandbox/bad.lsm'] [queued]
u000 (hrm@svi.nl): u000_ccc - ['tests/jobfiles/sandbox/bad.lsm'] [queued]
u111 (hrm@svi.nl): u111_fff - ['tests/jobfiles/sandbox/bad.lsm'] [queued]
u111 (hrm@svi.nl): u111_ggg - ['tests/jobfiles/sandbox/bad.lsm'] [queued]
========================= queue status =========================


******** removing jobs from queue: ********
jq.joblist: ['u000_aaa', 'u111_ddd', 'u000_bbb', 'u111_eee', 'u000_ccc', 'u111_fff', 'u111_ggg']
remove('u000_aaa'): u000_aaa (joblist: ['u000_bbb', 'u111_ddd', 'u000_ccc', 'u111_eee', 'u111_fff', 'u111_ggg'])
remove('u000_bbb'): u000_bbb (joblist: ['u000_ccc', 'u111_ddd', 'u111_eee', 'u111_fff', 'u111_ggg'])
next_job(): u000_ccc (joblist: ['u111_ddd', 'u111_eee', 'u111_fff', 'u111_ggg'])
remove('u111_ggg'): u111_ggg (joblist: ['u111_ddd', 'u111_eee', 'u111_fff'])
next_job(): u111_ddd (joblist: ['u111_eee', 'u111_fff'])
remove('u111_eee'): u111_eee (joblist: ['u111_fff'])
remove('u111_fff'): u111_fff (joblist: [])

----- no queued jobs should be left (but some processing) ----
jq.joblist: []
========================= queue status =========================
--- jobs retrieved for processing
u000 (hrm@svi.nl): u000_ccc - ['tests/jobfiles/sandbox/bad.lsm'] [queued]
u111 (hrm@svi.nl): u111_ddd - ['tests/jobfiles/sandbox/bad.lsm'] [queued]
------------------------- queue status -------------------------
--- jobs queued (not yet retrieved)
None.
========================= queue status =========================


******** trying to removing jobs from the empty queue: ********
None


******** trying to add duplicate jobs to queue: ********
jq.joblist: []
jq.joblist: ['u000_aaa']
Adding duplicate job failed as expected (Job with uid 'u000_aaa' already in this queue!)!
